### PR TITLE
fix: add runtime hints for trace

### PR DESCRIPTION
**Pull request by zhumin8**
~~**Builds on top of CI setup in #1933. Will rebase once its merged in.**~~

This pr:
- adds `TraceRuntimeHints` in trace autoconfig module and unit test for it. This allows native builds.
- adds `SampleRuntimeHints` this enables [pubsub trace](https://googlecloudplatform.github.io/spring-cloud-gcp//reference/html/index.html#pubsub-trace-instrumentation-experimental) portion of the sample to run after Native Build. not exactly sure why this is needed.
- adds `TestRuntimeHints` to allow logging portion of the test setup run as expect. (same as #1933)

CI test should run sample integration test.

To test run manually:
- run application, build with `mvn clean package -Pnative-sample-config -Pnative`, then run application with `./target/spring-cloud-gcp-trace-sample`. Then follow sample README to test functionality.
- run test: `mvn -PnativeTest clean test -Pnative-sample-config`
expect to get:
```
com.example.TraceSampleApplicationIntegrationTests > testTracesAreLoggedCorrectly() SUCCESSFUL


Test run finished after 18500 ms
[         3 containers found      ]
[         0 containers skipped    ]
[         3 containers started    ]
[         0 containers aborted    ]
[         3 containers successful ]
[         0 containers failed     ]
[         1 tests found           ]
[         0 tests skipped         ]
[         1 tests started         ]
[         0 tests aborted         ]
[         1 tests successful      ]
[         0 tests failed          ]

```

UPDATE: start failing on ci and local after rebase. Need investigate.

### CODE REVIEW

**Comment by mpeddada1** [comment]
Pasting in error message for reference: 
```
Failures (1):
  JUnit Jupiter:LoggingSampleApplicationIntegrationTests
    ClassSource [className = 'com.example.LoggingSampleApplicationIntegrationTests', filePosition = null]
    => java.lang.ExceptionInInitializerError
       org.springframework.test.context.TestContextManager.<init>(TestContextManager.java:113)
       org.junit.jupiter.engine.execution.ExtensionValuesStore.lambda$getOrComputeIfAbsent$4(ExtensionValuesStore.java:86)
       org.junit.jupiter.engine.execution.ExtensionValuesStore$MemoizingSupplier.computeValue(ExtensionValuesStore.java:223)
       org.junit.jupiter.engine.execution.ExtensionValuesStore$MemoizingSupplier.get(ExtensionValuesStore.java:211)
       org.junit.jupiter.engine.execution.ExtensionValuesStore$StoredValue.evaluate(ExtensionValuesStore.java:191)
       [...]
       Suppressed: java.lang.NoClassDefFoundError: Could not initialize class org.springframework.test.context.BootstrapUtils
         org.springframework.test.context.TestContextManager.<init>(TestContextManager.java:113)
         org.junit.jupiter.engine.execution.ExtensionValuesStore.lambda$getOrComputeIfAbsent$4(ExtensionValuesStore.java:86)
         org.junit.jupiter.engine.execution.ExtensionValuesStore$MemoizingSupplier.computeValue(ExtensionValuesStore.java:223)
         org.junit.jupiter.engine.execution.ExtensionValuesStore$MemoizingSupplier.get(ExtensionValuesStore.java:211)
         org.junit.jupiter.engine.execution.ExtensionValuesStore$StoredValue.evaluate(ExtensionValuesStore.java:191)
         [...]
     Caused by: java.lang.IllegalStateException: Failed to load class for @org.springframework.test.context.web.WebAppConfiguration
       org.springframework.test.context.BootstrapUtils.loadWebAppConfigurationClass(BootstrapUtils.java:213)
       org.springframework.test.context.BootstrapUtils.<clinit>(BootstrapUtils.java:63)
       [...]
     Caused by: java.lang.ClassNotFoundException: org.springframework.test.context.web.WebAppConfiguration
       java.base@17.0.5/java.lang.Class.forName(DynamicHub.java:1132)
       org.springframework.util.ClassUtils.forName(ClassUtils.java:284)
       org.springframework.test.context.BootstrapUtils.loadWebAppConfigurationClass(BootstrapUtils.java:209)
       [...]
```

**Comment by mpeddada1** [comment]
It looks like the argument to enable RunReachabilityHandlersConcurrently feature is still needed for this sample:
```
<plugin>
	<groupId>org.graalvm.buildtools</groupId>
	<artifactId>native-maven-plugin</artifactId>
	<configuration>
		<buildArgs>
		  <arg>-H:+RunReachabilityHandlersConcurrently</arg>
		</buildArgs>
	</configuration>
</plugin>
```
Without this argument, the tests fail at build time with the following error:

```
Failed generating 'native-tests' after 45.4s.

The build process encountered an unexpected error:

> java.lang.NullPointerException: Cannot invoke "java.util.Set.iterator()" because "triggers" is null
```

**Comment by mpeddada1** [comment]
Subdirectories for aot.factories don't get recognized. The trace sample fails with:
```
com.example.TraceSampleApplicationIntegrationTests > testTracesAreLoggedCorrectly() FAILED


Failures (1):
  JUnit Jupiter:TraceSampleApplicationIntegrationTests:testTracesAreLoggedCorrectly()
    MethodSource [className = 'com.example.TraceSampleApplicationIntegrationTests', methodName = 'testTracesAreLoggedCorrectly', methodParameterTypes = '']
    => com.oracle.svm.core.jdk.UnsupportedFeatureError: ThreadMXBean methods
       org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89)
       org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.management.SubstrateThreadMXBean.findDeadlockedThreads(SubstrateThreadMXBean.java:258)
       org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:157)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
       org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
       org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
       com.example.TraceSampleApplicationIntegrationTests.testTracesAreLoggedCorrectly(TraceSampleApplicationIntegrationTests.java:191)
       java.base@17.0.7/java.lang.reflect.Method.invoke(Method.java:568)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       [...]
```

File: spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/main/java/com/example/SampleRuntimeHints.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+public class SampleRuntimeHints implements RuntimeHintsRegistrar {
+
+  @Override
+  public void registerHints(RuntimeHints hints, ClassLoader classLoader) {
+    hints.reflection().registerTypes(
+        Arrays.asList(TypeReference.of(Application.class)),
+        hint -> hint.withMembers(
+            MemberCategory.INVOKE_PUBLIC_METHODS));
+  }
+}
```
Comment (mpeddada1) [inline spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/main/java/com/example/SampleRuntimeHints.java:34]: Without this configuration, the sample test results in:
```
com.example.TraceSampleApplicationIntegrationTests > testTracesAreLoggedCorrectly() FAILED


Failures (1):
  JUnit Jupiter:TraceSampleApplicationIntegrationTests:testTracesAreLoggedCorrectly()
    MethodSource [className = 'com.example.TraceSampleApplicationIntegrationTests', methodName = 'testTracesAreLoggedCorrectly', methodParameterTypes = '']
    => com.oracle.svm.core.jdk.UnsupportedFeatureError: ThreadMXBean methods
       org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89)
       org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.management.SubstrateThreadMXBean.findDeadlockedThreads(SubstrateThreadMXBean.java:258)
       org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:157)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
       org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
       org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
       com.example.TraceSampleApplicationIntegrationTests.testTracesAreLoggedCorrectly(TraceSampleApplicationIntegrationTests.java:191)
       java.base@17.0.7/java.lang.reflect.Method.invoke(Method.java:568)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       [...]

```

File: spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/aot/TraceRuntimeHints.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    Method method =
+        ReflectionUtils.findMethod(
+            ObservedAspect.class, "observeMethod", ProceedingJoinPoint.class);
+    hints
+        .reflection()
+        .registerMethod(method, ExecutableMode.INVOKE)
+        .registerTypes(
+            Arrays.asList(
+                TypeReference.of(com.google.protobuf.Value.class),
+                TypeReference.of(com.google.protobuf.Value.Builder.class)),
```
[the remark this answers is not in the document]
Comment (burkedavison) [reply spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/aot/TraceRuntimeHints.java:36]: I don't see `com.google.protobuf.Value` as an entry in the Trace library's reflect-config.json. It does show up for some, though: https://github.com/search?q=repo%3Agoogleapis%2Fgoogle-cloud-java+com.google.protobuf.Value+path%3A**%2Freflect-config.json&type=code&p=1

- [ ] Do we need to adjust the generation logic to include more than messages defined in the protos, or is a non-trace library triggering the reflection lookup?

**Review by burkedavison — APPROVED** [review]

**Comment by mpeddada1** [comment]
Thanks for the review @burkedavison! Adding some initial findings here. The dependency tree of the sample shows that the protobuf-java dependency which contains the `com.google.protobuf.Value` is brought in from google-cloud-trace:
```
[INFO] +- com.google.api.grpc:grpc-google-cloud-trace-v1:jar:2.20.0:test
[INFO] |  +- io.grpc:grpc-api:jar:1.55.1:compile
[INFO] |  +- io.grpc:grpc-context:jar:1.55.1:compile
[INFO] |  +- com.google.code.findbugs:jsr305:jar:3.0.2:compile
[INFO] |  +- com.google.errorprone:error_prone_annotations:jar:2.18.0:compile
[INFO] |  +- io.grpc:grpc-stub:jar:1.55.1:compile
[INFO] |  +- io.grpc:grpc-protobuf:jar:1.55.1:compile
[INFO] |  +- com.google.api.grpc:proto-google-common-protos:jar:2.21.0:compile
[INFO] |  +- io.grpc:grpc-protobuf-lite:jar:1.55.1:compile
[INFO] |  +- com.google.protobuf:protobuf-java:jar:3.23.2:compile
[INFO] |  +- com.google.api.grpc:proto-google-cloud-trace-v1:jar:2.20.0:test
[INFO] |  +- com.google.guava:guava:jar:31.1-jre:compile
[INFO] |  +- com.google.guava:failureaccess:jar:1.0.1:compile
[INFO] |  +- com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava:compile
[INFO] |  +- org.checkerframework:checker-qual:jar:3.33.0:compile
[INFO] |  \- com.google.j2objc:j2objc-annotations:jar:2.8:compile
```
Digging deeper to see if the stacktrace gives us more clues on how this class is being invoked.

File: spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/aot/TraceRuntimeHints.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mpeddada1) [reply spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/aot/TraceRuntimeHints.java:36]: Subdirectories for aot.factories don't get recognized. The trace sample fails with:
```
com.example.TraceSampleApplicationIntegrationTests > testTracesAreLoggedCorrectly() FAILED


Failures (1):
  JUnit Jupiter:TraceSampleApplicationIntegrationTests:testTracesAreLoggedCorrectly()
    MethodSource [className = 'com.example.TraceSampleApplicationIntegrationTests', methodName = 'testTracesAreLoggedCorrectly', methodParameterTypes = '']
    => com.oracle.svm.core.jdk.UnsupportedFeatureError: ThreadMXBean methods
       org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89)
       org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.management.SubstrateThreadMXBean.findDeadlockedThreads(SubstrateThreadMXBean.java:258)
       org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:157)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
       org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
       org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
       com.example.TraceSampleApplicationIntegrationTests.testTracesAreLoggedCorrectly(TraceSampleApplicationIntegrationTests.java:191)
       java.base@17.0.7/java.lang.reflect.Method.invoke(Method.java:568)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       [...]
```

Using ImportRuntimeHints annotation instead to resolve the failure mentioned in https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/1990#issuecomment-1632916647

**Comment by mpeddada1** [comment]
Subdirectories for aot.factories don't get recognized. The trace sample fails with:
```
com.example.TraceSampleApplicationIntegrationTests > testTracesAreLoggedCorrectly() FAILED


Failures (1):
  JUnit Jupiter:TraceSampleApplicationIntegrationTests:testTracesAreLoggedCorrectly()
    MethodSource [className = 'com.example.TraceSampleApplicationIntegrationTests', methodName = 'testTracesAreLoggedCorrectly', methodParameterTypes = '']
    => com.oracle.svm.core.jdk.UnsupportedFeatureError: ThreadMXBean methods
       org.graalvm.nativeimage.builder/com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:89)
       org.graalvm.nativeimage.builder/com.oracle.svm.core.jdk.management.SubstrateThreadMXBean.findDeadlockedThreads(SubstrateThreadMXBean.java:258)
       org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:157)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
       org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
       org.awaitility.core.ConditionFactory.until(ConditionFactory.java:985)
       org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:769)
       com.example.TraceSampleApplicationIntegrationTests.testTracesAreLoggedCorrectly(TraceSampleApplicationIntegrationTests.java:191)
       java.base@17.0.7/java.lang.reflect.Method.invoke(Method.java:568)
       org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:727)
       [...]
```

Using ImportRuntimeHints annotation instead to resolve the failure mentioned in https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/1990#issuecomment-1632916647

File: spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/aot/TraceRuntimeHints.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mpeddada1) [reply spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/aot/TraceRuntimeHints.java]: Good point! Analyzing the stacktrace of the test shows that the use of `com.google.protobuf.Value` is coming from https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/154cbe5c7a4f3be6eba76e6607ae5fa0528bf909/spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/src/test/java/com/example/TraceSampleApplicationIntegrationTests.java#L238

This line of code eventually leads to `GeneratedMessageV3#getMethodOrDie` being called:
Screenshot 2023-07-31 at 6 07 51 PM

[the remark this answers is not in the document]
Comment (mpeddada1) [reply spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/trace/aot/TraceRuntimeHints.java]: [Update] Tests are passing without these configurations after rebasing with `origin/main`.

**Pull request merged**